File: src/algorithms/sets/membership/count-min-sketch/sources/count-min-sketch.py
"""
Count-Min Sketch — probabilistic frequency estimation using a d×w counter matrix.
Supports sub-linear space frequency estimation with one-sided error (never undercounts).
Time: O(d) per insert/query — Space: O(d × w)
"""

from typing import List, Dict


def compute_sketch_hash(value: int, hash_idx: int, width: int) -> int:
    return abs((value * (hash_idx * 1327 + 31) + hash_idx * 7919) % width)  # @step:hash-element


def count_min_sketch(
    elements: List[int],
    queries: List[int],
    width: int,
    depth: int,
) -> Dict[str, List[Dict[str, int]]]:
    # Initialize d×w counter matrix with all zeros
    sketch = [[0] * width for _ in range(depth)]  # @step:initialize

    # Insert phase: for each element, increment d counters
    for element in elements:
        for hash_idx in range(depth):
            col = compute_sketch_hash(element, hash_idx, width)
            sketch[hash_idx][col] += 1  # @step:increment-count

    # Query phase: estimate frequency by taking minimum across all d rows
    results = []
    for query in queries:
        min_count = min(  # @step:check-membership
            sketch[hash_idx][compute_sketch_hash(query, hash_idx, width)]
            for hash_idx in range(depth)
        )
        if min_count > 0:
            results.append({"value": query, "estimatedCount": min_count})  # @step:member-found
        else:
            pass  # @step:member-not-found

    return {"results": results}  # @step:complete

File: src/algorithms/sets/membership/count-min-sketch/sources/count-min-sketch.go
// Count-Min Sketch — probabilistic frequency estimation using a d×w counter matrix.
// Supports sub-linear space frequency estimation with one-sided error (never undercounts).
// Time: O(d) per insert/query — Space: O(d × w)

package main

import (
	"fmt"
	"math"
)

func computeSketchHash(value int, hashIdx int, width int) int {
	result := int(math.Abs(float64((value*(hashIdx*1327+31) + hashIdx*7919) % width)))
	return result // @step:hash-element
}

type EstimatedResult struct {
	value         int
	estimatedCount int
}

func countMinSketch(elements []int, queries []int, width int, depth int) []EstimatedResult {
	// Initialize d×w counter matrix with all zeros
	sketch := make([][]int, depth)
	for rowIdx := range sketch {
		sketch[rowIdx] = make([]int, width)
	}
	// @step:initialize

	// Insert phase: for each element, increment d counters
	for _, element := range elements {
		for hashIdx := 0; hashIdx < depth; hashIdx++ {
			col := computeSketchHash(element, hashIdx, width)
			sketch[hashIdx][col]++ // @step:increment-count
		}
	}

	// Query phase: estimate frequency by taking minimum across all d rows
	results := make([]EstimatedResult, 0)
	for _, query := range queries {
		minCount := math.MaxInt // @step:check-membership
		for hashIdx := 0; hashIdx < depth; hashIdx++ {
			col := computeSketchHash(query, hashIdx, width)
			if sketch[hashIdx][col] < minCount {
				minCount = sketch[hashIdx][col]
			}
		}
		estimatedCount := 0
		if minCount != math.MaxInt {
			estimatedCount = minCount
		}
		if estimatedCount > 0 {
			results = append(results, EstimatedResult{query, estimatedCount}) // @step:member-found
		} else {
			_ = query // @step:member-not-found
		}
	}

	return results // @step:complete
}

func main() {
	elements := []int{1, 2, 1, 3, 2, 1}
	queries := []int{1, 2, 3, 4}
	results := countMinSketch(elements, queries, 10, 3)
	for _, result := range results {
		fmt.Printf("value=%d count=%d\n", result.value, result.estimatedCount)
	}
}
